openshift-client/src/main/java/io/fabric8/openshift/client/internal/OpenShiftOAuthInterceptor.java

```diff
@@ -16,124 +16,130 @@
+  }
 
-    @Override
-    public Response intercept(Chain chain) throws IOException {
-        Request request = chain.request();
+  @Override
+  public Response intercept(Chain chain) throws IOException {
+    Request request = chain.request();
 
-        //Build new request
-        Request.Builder builder = request.newBuilder();
-        String token = null;
+    //Build new request
+    Request.Builder builder = request.newBuilder();
 
-        if (Utils.isNotNullOrEmpty(oauthToken.get()) && Utils.isNullOrEmpty(request.header(AUTHORIZATION))) {
-          token = oauthToken.get();
-        } else if (Utils.isNotNullOrEmpty(config.getUsername()) && Utils.isNotNullOrEmpty(config.getPassword())) {
-          token = acquireNewToken();
-        } else if (Utils.isNotNullOrEmpty(config.getOauthToken())) {
-          token = config.getOauthToken();
-          oauthToken.set(token);
-        }
-
-        // avoid overwriting basic auth token with stale bearer token
-        if (Utils.isNotNullOrEmpty(token) && Utils.isNullOrEmpty(request.header(AUTHORIZATION))) {
-            setAuthHeader(builder, token);
-        }
+    String token = oauthToken.get();
+    // avoid overwriting basic auth token with stale bearer token
+    if (Utils.isNotNullOrEmpty(token) && Utils.isNullOrEmpty(request.header(AUTHORIZATION))) {
+      setAuthHeader(builder, token);
+    }
 
-        request = builder.build();
-        Response response = chain.proceed(request);
+    request = builder.build();
+    Response response = chain.proceed(request);
 
-        //If response is Forbidden or Unauthorized, try to obtain a token via authorize() or via config.
-        if (response.code() != 401 && response.code() != 403) {
-          return response;
-        } else if (Utils.isNotNullOrEmpty(config.getUsername()) && Utils.isNotNullOrEmpty(config.getPassword())) {
-          token = acquireNewToken();
-        } else if (Utils.isNotNullOrEmpty(config.getOauthToken())) {
-          token = config.getOauthToken();
+    //If response is Forbidden or Unauthorized, try to obtain a token via authorize() or via config.
+    if (response.code() != 401 && response.code() != 403) {
+      return response;
+    } else if (Utils.isNotNullOrEmpty(config.getUsername()) && Utils.isNotNullOrEmpty(config.getPassword())) {
+      synchronized (client) {
+        // current token (if exists) is borked, don't resend
+        oauthToken.set(null);
+        token = authorize();
+        if (token != null) {
           oauthToken.set(token);
         }
+      }
+    } else if (Utils.isNotNullOrEmpty(config.getOauthToken())) {
+      token = config.getOauthToken();
+      oauthToken.set(token);
+    }
 
 
-      //If token was obtain, then retry request using the obtained token.
-      if (Utils.isNotNullOrEmpty(token)) {
-        // Close the previous response to prevent leaked connections.
-        response.body().close();
+    //If token was obtain, then retry request using the obtained token.
```

Comment:
I know you didn't make this change; might be wrong as I'm not native English speaker, but shouldn't it be obtained here?

Reply:
Yes, You're right. Fixing that